### PR TITLE
Avoid mutex ordering issues.

### PR DESCRIPTION
**Pull request by bbockelm**
Prevent m_source_mutex from being held when OpenHandler::open is called
(which will then take the internal m_mutex).  This avoids a lock re-ordering
issue.

### CODE REVIEW

**Comment by slava77** [comment]
@bbockelm how serious is this one?
Is it worth to make another release or just wait until other things accumulate in the next few days.

**Comment by bbockelm** [comment]
It can likely wait.  @Dr15Jones noticed this in the relvals - we see no reason why it couldn't trigger in the T0, but the likelihood of hitting the race is smaller than the other XrdAdaptor fix.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@bbockelm 
we will also need 76X/80X PR(s)

File: Utilities/XrdAdaptor/src/XrdRequestManager.cc [hunk 1]
```diff
@@ -370,10 +370,14 @@ RequestManager::checkSourcesImpl(timespec &now, IOSize requestSize)
...
   }
+  // NOTE: OpenHandler will take an internal mutex, then invoke RequestManager methods that take
+  // m_source_mutex.  This ordering is also done within OpenHandler::HandleResponseWithHosts; hence,
+  // we must release m_source_mutex before calling open() or deadlock due to mutex ordering.
   if (findNewSource)
   {
-    m_open_handler->open();
     m_lastSourceCheck = now;
+    sentry.unlock();
+    m_open_handler->open();
```
Comment (Dr15Jones) [inline Utilities/XrdAdaptor/src/XrdRequestManager.cc]: need to reacquire the lock after this line.

**Comment by bbockelm** [comment]
Fixed @Dr15Jones code comment.

I'll open PRs for the other branches once this goes in (and we can cherrypick).

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
I'll see if I can use this change on jobs which have been failing

File: Utilities/XrdAdaptor/src/XrdRequestManager.cc [hunk 2]
```diff
@@ -370,10 +370,16 @@ RequestManager::checkSourcesImpl(timespec &now, IOSize requestSize)
...
+  // m_source_mutex.  This ordering is also done within OpenHandler::HandleResponseWithHosts; hence,
+  // we must release m_source_mutex before calling open() or deadlock due to mutex ordering.
   if (findNewSource)
   {
-    m_open_handler->open();
     m_lastSourceCheck = now;
+    sentry.unlock();
+    // Automatically reacquire the lock after this code block finishes.
+    std::shared_ptr<const void> reacquire_lock_sentry(nullptr, [&sentry](void*) {sentry.lock();});
+    m_open_handler->open();
```
Comment (Dr15Jones) [inline Utilities/XrdAdaptor/src/XrdRequestManager.cc]: Turns out this is not safe. There is a comment in OpenHandler::open() which explicitly states that this function isn't thread safe and requires a lock on it to guarantee you can't call it from multiple threads. The problem is an unguarded access to a member data.

**Comment by Dr15Jones** [comment]
@bbockelm What's next try for this fix?

**Comment by Dr15Jones** [comment]
@bbockelm Should we go with just one mutex?

**Comment by bbockelm** [comment]
Nope - it turns out the fix is more difficult than that.

Basically, depending on the context, the callback handler may-or-may-not be invoked with a Xrootd mutex held (I really, really hate this fact).  This same mutex will be needed for XrdFile::open.

So, any mutex held within the callback MUST NOT be held when we call XrdFile::open.  This is why we unlocked `m_mutex` before; we don't currently have this problem with `m_source_mutex` because of the deadlock that triggered this ticket.  If we merge the two together, we'll bring back the old issue.

Rather, I think we need to get rid of the use of `m_source_mutex` in the callback.  It is protecting read-only data that changes whenever a server throws an error - so, on the order of minutes.  Further, it is only used to format error messages.

I am working to pre-format the parts of the error message that require `m_source_mutex` and caching them in an lock-free manner.  This removes the use of `m_source_mutex` in the callback function, allowing us to keep it held when we call `XrdFile::open`.

**Comment by davidlange6** [comment]
Hi - I'm cleaning up the 75x CMSSW release queue, so closing issues not directly related to analysis or finalizing the HI MC production. Please make sure these requests are included in the 80x release already. Thanks!

**Pull request closed without merging**